File: chatbot/core/ai_data_loader.py
"""
Система загрузки данных значков из ai-data структуры
"""
import json
import os
from typing import Dict, List, Optional
from pathlib import Path

from models.badge import Badge, BadgeLevel, Category, BadgeData


class AIDataLoader:
    """Загрузчик данных значков из ai-data структуры"""
    
    def __init__(self, ai_data_path: str = "../ai-data"):
        """
        Инициализация загрузчика
        
        Args:
            ai_data_path: Путь к папке ai-data
        """
        self.ai_data_path = Path(ai_data_path)
        self._badge_data: Optional[BadgeData] = None
        self._categories_cache: Dict[str, Category] = {}
        self._badges_cache: Dict[str, Badge] = {}
    
    def load_all_data(self) -> BadgeData:
        """Загружает все данные значков из ai-data структуры"""
        if self._badge_data is not None:
            return self._badge_data
        
        if not self.ai_data_path.exists():
            raise FileNotFoundError(f"Не найдена папка {self.ai_data_path}")
        
        # Загружаем главный индекс
        master_index_path = self.ai_data_path / "MASTER_INDEX.json"
        if not master_index_path.exists():
            raise FileNotFoundError(f"Не найден файл {master_index_path}")
        
        with open(master_index_path, 'r', encoding='utf-8') as f:
            master_index = json.load(f)
        
        # Загружаем все категории
        categories = []
        for category_info in master_index["categories"]:
            category = self._load_category_from_ai_data(category_info)
            categories.append(category)
        
        # Создаем объект BadgeData
        self._badge_data = BadgeData(
            project="Путеводитель",
            version="1.0",
            totalCategories=master_index["totalCategories"],
            totalBadges=master_index["totalBadges"],
            categories=categories
        )
        
        return self._badge_data
    
    def _load_category_from_ai_data(self, category_info: dict) -> Category:
        """Загружает категорию из ai-data структуры"""
        category_id = category_info["id"]
        
        # Загружаем индекс категории
        category_path = self.ai_data_path / category_info["path"]
        index_path = category_path / "index.json"
        
        if not index_path.exists():
            raise FileNotFoundError(f"Не найден файл {index_path}")
        
        with open(index_path, 'r', encoding='utf-8') as f:
            category_index = json.load(f)
        
        # Загружаем значки категории
        badges = []
        for badge_info in category_index.get("badgesData", []):
            badge = self._load_badge_from_ai_data(category_path, badge_info)
            badges.append(badge)
        
        # Создаем категорию
        category = Category(
            id=category_id,
            title=category_info["title"],
            emoji=category_info["emoji"],
            path=category_info["path"],
            badges=badges
        )
        
        return category
    
    def _load_badge_from_ai_data(self, category_path: Path, badge_info: dict) -> Badge:
        """Загружает значок из ai-data структуры"""
        badge_id = badge_info["id"]
        
        # Загружаем файл значка
        badge_file = category_path / f"{badge_id}.json"
        if not badge_file.exists():
            raise FileNotFoundError(f"Не найден файл {badge_file}")
        
        with open(badge_file, 'r', encoding='utf-8') as f:
            badge_data = json.load(f)
        
        # Загружаем уровни значка
        levels = []
        for level_data in badge_data.get("levels", []):
            level = BadgeLevel(
                id=level_data["id"],
                level=level_data["level"],
                title=level_data["title"],
                emoji=level_data["emoji"],
                criteria=level_data.get("criteria", ""),
                confirmation=level_data.get("confirmation", "")
            )
            levels.append(level)
        
        # Создаем значок
        badge = Badge(
            id=badge_id,
            title=badge_data["title"],
            emoji=badge_data["emoji"],
            categoryId=badge_data["categoryId"],
            description=badge_data.get("description", ""),
            skillTips=badge_data.get("skillTips", ""),
            examples=badge_data.get("examples", ""),
            philosophy=badge_data.get("philosophy", ""),
            howToBecome=badge_data.get("howToBecome", ""),
            nameExplanation=badge_data.get("nameExplanation", ""),
            levels=levels
        )
        
        return badge
    
    def get_category_by_id(self, category_id: str) -> Optional[Category]:
        """Получает категорию по ID"""
        if not self._badge_data:
            self.load_all_data()
        
        for category in self._badge_data.categories:
            if category.id == category_id:
                return category
        return None
    
    def get_badge_by_id(self, badge_id: str) -> Optional[Badge]:
        """Получает значок по ID"""
        if not self._badge_data:
            self.load_all_data()
        
        for category in self._badge_data.categories:
            for badge in category.badges:
                if badge.id == badge_id:
                    return badge
        return None
    
    def search_badges(self, query: str) -> List[Badge]:
        """Поиск значков по запросу"""
        if not self._badge_data:
            self.load_all_data()
        
        query_lower = query.lower()
        results = []
        
        for category in self._badge_data.categories:
            for badge in category.badges:
                if (query_lower in badge.title.lower() or 
                    query_lower in badge.description.lower() or
                    query_lower in badge.importance.lower()):
                    results.append(badge)
        
        return results

File: cf-api/src/neurovalyusha/ai_data_loader.ts
import type { Fetcher, KVNamespace } from '@cloudflare/workers-types'
import { kvGetJson, kvPutJson } from './kv'

export type BadgeFields = 'minimal' | 'standard' | 'full'

export type MasterIndexCategory = {
  id: string
  title?: string
  emoji?: string
  path?: string
}

export type MasterIndex = {
  project?: string
  version?: string
  categories?: MasterIndexCategory[]
}

export type CategoryIndexBadge = {
  id: string
  title?: string
  emoji?: string
  levels?: string[]
}

export type CategoryIndex = {
  categoryId?: string
  title?: string
  badgesData?: CategoryIndexBadge[]
}

export type BadgeLevel = {
  id?: string
  level?: string | number
  title?: string
  emoji?: string
  criteria?: string | string[]
  confirmation?: string | string[]
}

export type BadgeData = {
  id: string
  title?: string
  emoji?: string
  categoryId?: string
  categoryTitle?: string
  description?: string
  skillTips?: string
  nameExplanation?: string
  examples?: string | string[]
  philosophy?: string
  howToBecome?: string
  importance?: string
  levels?: BadgeLevel[]
}

type AiDataEnv = {
  ASSETS?: Fetcher
  NEUROVALYUSHA_KV?: KVNamespace
}

const DEFAULT_TTL_SECONDS = 60 * 10

export function clip(text: string, max: number): string {
  const cleaned = String(text || '').trim()
  if (!cleaned) return ''
  if (cleaned.length <= max) return cleaned
  return `${cleaned.slice(0, max).trim()}...`
}

export async function loadMasterIndex(env: AiDataEnv, params?: { maxAgeSeconds?: number }): Promise<MasterIndex | null> {
  const ttlSeconds = params?.maxAgeSeconds ?? DEFAULT_TTL_SECONDS
  const kv = env.NEUROVALYUSHA_KV
  const cacheKey = 'nv:ai-data:master'

  const cached = await kvGetJson<MasterIndex>(kv, cacheKey)
  if (cached) return cached

  const res = await fetchAsset(env, '/ai-data/MASTER_INDEX.json')
  if (!res.ok) return cached ?? null
  const data = (await res.json().catch(() => null)) as MasterIndex | null
  if (!data || typeof data !== 'object') return cached ?? null

  await kvPutJson(kv, cacheKey, data, { ttlSeconds })
  return data
}

export async function loadCategoryIndex(
  env: AiDataEnv,
  categoryId: string,
  params?: { maxAgeSeconds?: number },
): Promise<CategoryIndex | null> {
  const ttlSeconds = params?.maxAgeSeconds ?? DEFAULT_TTL_SECONDS
  const kv = env.NEUROVALYUSHA_KV
  const normalizedId = String(categoryId || '').trim()
  if (!normalizedId) return null

  const cacheKey = `nv:ai-data:category:${normalizedId}`
  const cached = await kvGetJson<CategoryIndex>(kv, cacheKey)
  if (cached) return cached

  const res = await fetchAsset(env, `/ai-data/category-${normalizedId}/index.json`)
  if (!res.ok) return cached ?? null
  const data = (await res.json().catch(() => null)) as CategoryIndex | null
  if (!data || typeof data !== 'object') return cached ?? null

  await kvPutJson(kv, cacheKey, data, { ttlSeconds })
  return data
}

export async function loadBadgeData(
  env: AiDataEnv,
  badgeId: string,
  fields: BadgeFields = 'standard',
  params?: { maxAgeSeconds?: number },
): Promise<BadgeData | null> {
  const ttlSeconds = params?.maxAgeSeconds ?? DEFAULT_TTL_SECONDS
  const kv = env.NEUROVALYUSHA_KV
  const normalizedId = normalizeBadgeId(badgeId)
  if (!normalizedId) return null

  const cacheKey = `nv:ai-data:badge:${normalizedId}:${fields}`
  const cached = await kvGetJson<BadgeData>(kv, cacheKey)
  if (cached) return cached

  const categoryId = normalizedId.split('.')[0]
  if (!categoryId) return cached ?? null

  const res = await fetchAsset(env, `/ai-data/category-${categoryId}/${normalizedId}.json`)
  if (!res.ok) return cached ?? null

  const raw = (await res.json().catch(() => null)) as BadgeData | null
  if (!raw || typeof raw !== 'object' || !raw.id) return cached ?? null

  const categoryTitle = (await loadCategoryIndex(env, raw.categoryId || categoryId))?.title
  const withCategory: BadgeData = { ...raw, categoryId: raw.categoryId || categoryId, categoryTitle }

  const shaped = applyBadgeFields(withCategory, fields)
  await kvPutJson(kv, cacheKey, shaped, { ttlSeconds })
  return shaped
}

function normalizeBadgeId(raw: string): string {
  const text = String(raw || '').trim()
  if (!text) return ''
  const match = text.match(/\d{1,2}\.\d{1,2}/)
  return match ? match[0] : text
}

function applyBadgeFields(badge: BadgeData, fields: BadgeFields): BadgeData {
  if (fields === 'full') return badge

  const minimal: BadgeData = {
    id: badge.id,
    title: badge.title,
    emoji: badge.emoji,
    categoryId: badge.categoryId,
    categoryTitle: badge.categoryTitle,
    description: clipOptionalText(badge.description, 140),
    skillTips: clipOptionalText(badge.skillTips, 200),
  }

  if (fields === 'minimal') return minimal

  return {
    ...minimal,
    nameExplanation: badge.nameExplanation,
    importance: badge.importance,
    philosophy: badge.philosophy,
    howToBecome: badge.howToBecome,
    examples: clipOptionalText(badge.examples, 100),
    levels: badge.levels,
  }
}

function clipOptionalText(value: unknown, max: number): string | undefined {
  const text = coerceText(value)
  if (!text) return undefined
  const clipped = clip(text, max)
  return clipped || undefined
}

function coerceText(value: unknown): string {
  if (Array.isArray(value)) {
    return value.filter((item) => typeof item === 'string' && item.trim().length > 0).join('\n')
  }
  if (typeof value === 'string') return value
  return ''
}

async function fetchAsset(env: AiDataEnv, pathname: string): Promise<Response> {
  if (env.ASSETS) {
    return await env.ASSETS.fetch(new Request(`http://localhost${pathname}`))
  }
  return await fetch(pathname)
}
